Migrate pipelines to 1ES

habbes opened the pull request:
This PR migrates the build pipeline to 1ES.

I've removed the `Debug` build stage to save resources because we're exceeding the 60 min runtime limit when running both jobs.

I've edited the hardcoded fallback path of the "enlistement" root to match the working directory of the new build agents. This fallback path is reached when we can't find the correct path based on the current working directory.

gathogojr commented on test/FunctionalTests/Framework/Util/FrameworkUtil.cs:
```diff
@@ -191,15 +200,17 @@ private static string FindEnlistmentRootHeuristically()
-            // We check the result path again to ensure it is an ENLISTMENT_ROOT.
+
+
+            //// We check the result path again to ensure it is an ENLISTMENT_ROOT.
```
nit:
```suggestion
            // We check the result path again to ensure it is an ENLISTMENT_ROOT.
```

gathogojr commented on test/FunctionalTests/Framework/Util/FrameworkUtil.cs:
```diff
@@ -191,15 +200,17 @@ private static string FindEnlistmentRootHeuristically()
             }
 
-            
+
```
Do we need extra empty lines above and below this one?

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+- name: __NugetSecurityAnalysisWarningLevel__
+  value: none
+- name: BuildConfiguration
+  value: release
+- name: BuildPlatform
```
These two variables, do you need them here given that you have both a `Debug` and `Release` job where you specify the respective build configuration?

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+variables:
+- name: __NugetSecurityAnalysisWarningLevel__
+  value: none
+- name: BuildConfiguration
+  value: release
```
These two variables, do you need them here given that you have both a Debug and Release job where you specify the respective build configuration?

habbes replied:
Removed.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+  value: release
+- name: BuildPlatform
+  value: anycpu
+- name: ProductBinPath
+  value: $(Build.SourcesDirectory)\bin\$(BuildPlatform)\$(BuildConfiguration)
```
Do you use this variable anywhere?

habbes replied:
No. Removed.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+      - job:  Debug	
+        timeoutInMinutes: 120
+        variables:	
+            buildPlatform: 'Any CPU'	
```
Is this `buildPlatform` (lower-camel-case) variable different from the `BuildPlatform` specify at the outer scope?
Why is it that the one at the outer scope is assigned the value "anycpu" (no space) while in this scope it's assigned the value "Any CPU"?

habbes replied:
The inner scope overrides the outer scope. The change in casing must be an artifact of the conversion tooling. I've harmonized the variables and removed duplicates.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+        timeoutInMinutes: 120
+        variables:	
+            buildPlatform: 'Any CPU'	
+            buildConfiguration: 'Debug'	
```
Is this `buildConfiguration` (lower-camel-case) variable different from the `BuildConfiguration` specify at the outer scope?

habbes replied:
I've removed the duplicate variables.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+            mainDll3: 'Microsoft.Spatial.dll'	
+            ProductBinPath4: '$(Build.SourcesDirectory)\src\Microsoft.OData.Client\bin\$(BuildConfiguration)'	
+            mainDll4: 'Microsoft.OData.Client.dll'	
+            ProductBinPath5: '$(Build.SourcesDirectory)\src\CodeGen\bin\$(BuildConfiguration)'	
+            mainDll5: 'Microsoft.OData.Service.Design.T4.dll'	
```
Which one is this one? Could it be legacy leftover that we can safely get rid off?

habbes replied:
I just checked and it's no longer being used. Must be a legacy item. Removed.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+            mainDll4: 'Microsoft.OData.Client.dll'	
+            ProductBinPath5: '$(Build.SourcesDirectory)\src\CodeGen\bin\$(BuildConfiguration)'	
+            mainDll5: 'Microsoft.OData.Service.Design.T4.dll'	
+            ProductBinPath: $(Build.SourcesDirectory)\bin\$(BuildPlatform)\$(BuildConfiguration)\Product	
```
Are we redefining `ProductBinPath` defined in the outer scope? What is this variable used for?

habbes replied:
I've just checked and it's no longer being used. I've removed it.

gathogojr commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+            artifactName: NuGetDebug	
+        steps:	
+          - template: /nightly.yml@self
+      - job:  Release
```
The requests I raised in the `Debug` job section are relevant here as well:

- https://github.com/OData/odata.net/pull/2965/files#r1607127969
- https://github.com/OData/odata.net/pull/2965/files#r1607130100
- https://github.com/OData/odata.net/pull/2965/files#r1607132058
- https://github.com/OData/odata.net/pull/2965/files#r1607133198

habbes replied:
I've applied the changes here as well.

gathogojr commented on azure-pipelines-rolling.yml:
```diff
@@ -12,55 +12,83 @@ trigger:
+    stages:
+    - stage: build
+      jobs:
+      - job:  Debug	
```
The requests I raised in the `Debug` job section for nightly pipeline are relevant here as well:

- https://github.com/OData/odata.net/pull/2965/files#r1607127969
- https://github.com/OData/odata.net/pull/2965/files#r1607130100
- https://github.com/OData/odata.net/pull/2965/files#r1607132058
- https://github.com/OData/odata.net/pull/2965/files#r1607133198

habbes replied:
I've applied the changes here as well.

gathogojr commented on azure-pipelines-rolling.yml:
```diff
@@ -12,55 +12,83 @@ trigger:
+        steps:	
+          - template: /buildandtest.yml@self	
+          - template: /credscan.yml@self
+      - job:  Release
```
The requests I raised in the `Debug` job section for nightly pipeline are relevant here as well:

- https://github.com/OData/odata.net/pull/2965/files#r1607127969
- https://github.com/OData/odata.net/pull/2965/files#r1607130100
- https://github.com/OData/odata.net/pull/2965/files#r1607132058
- https://github.com/OData/odata.net/pull/2965/files#r1607133198

habbes replied:
I've applied the changes here as well.

gathogojr commented on nightly.yml:
```diff
@@ -10,10 +10,10 @@
-      displayName: 'Publish SBOM'
-      inputs:
-        PathtoPublish: '$(Build.ArtifactStagingDirectory)\$(NugetArtifactsDir)\sbom'
+    # - task: 1ES.PublishPipelineArtifact@1
+    #   displayName: 'Publish SBOM'
+    #   inputs:
+    #     path: '$(Build.ArtifactStagingDirectory)\$(NugetArtifactsDir)\sbom'
```
Do we not need this any more?
If we don't, shouldn't we get rid of the task instead of commenting it out?

gathogojr commented on nightly.yml:
```diff
@@ -175,17 +175,9 @@
-      inputs:
-        PathtoPublish: '$(Build.ArtifactStagingDirectory)\$(NugetArtifactsDir)'
-        ArtifactName: $(NugetArtifactsDir)
+    # - task: 1ES.PublishPipelineArtifact@1
+    #   displayName: 'Publish Artifact - Nuget Packages'
+    #   inputs:
+    #     path: '$(Build.ArtifactStagingDirectory)\$(NugetArtifactsDir)'
+    #     artifact: $(NugetArtifactsDir)
```
Why comment this out instead of getting rid of it?

gathogojr commented on test/FunctionalTests/Framework/Util/FrameworkUtil.cs:
```diff
@@ -164,8 +164,17 @@ private static string FindEnlistmentRootHeuristically()
             }
 
-            // Second use current assembly location to determine.
+            // Second, use environment variable defined in build agents
+            // See: https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables
+            result = Environment.GetEnvironmentVariable("BUILD_SOURCESDIRECTORY");
+            if (!string.IsNullOrEmpty(result))
+            {
+                return result;
+            }
```
It's unlikely that we're going to revert to the AzDO pipelines any time soon. I think you should move this block to the top of the method

gathogojr commented on nightly.yml:
```diff
@@ -10,10 +10,10 @@
+    # - task: 1ES.PublishPipelineArtifact@1
+    #   displayName: 'Publish SBOM'
+    #   inputs:
+    #     path: '$(Build.ArtifactStagingDirectory)\$(NugetArtifactsDir)\sbom'
```
We generate the manifest in the previous 2 steps. If we're getting rid of this step, do we need to retain those 2 other steps?

gathogojr replied:
If we got rid of the other 2 steps, will there be any value in retaining this _nightly.yml_ file?

habbes replied:
The nightly also performs code signing

habbes commented on azure-pipelines-nightly.yml:
```diff
@@ -14,55 +14,103 @@ schedules:
+            mainDll4: 'Microsoft.OData.Client.dll'	
+            ProductBinPath5: '$(Build.SourcesDirectory)\src\CodeGen\bin\$(BuildConfiguration)'	
+            mainDll5: 'Microsoft.OData.Service.Design.T4.dll'	
+            ProductBinPath: $(Build.SourcesDirectory)\bin\$(BuildPlatform)\$(BuildConfiguration)\Product	
```
I've removed this variable.

habbes commented:
@gathogojr following your feedback I've done some cleanup:
- removed redundant or unused variables
- renamed all custom variables using camel-case for consistency
- renamed the ProductBinPath<1|2|3|4> and mainDll<1|2|3|4>variables to more meaningful names like `productBinPathEdm`, `mainDllSpatial`, etc.

gathogojr reviewed (approved):
LGTM

ElizabethOkerio commented:
Can the two failing runs be removed or what does it mean? Also, why do we have 3 runs for rolling, nightly etc instead of just the build and release?
